README.md
![alt](bin/SSicon.png)
SubString v 1.1
===============

The SubString package consists of three separate command-line modules (substring-A, substring-B and an auxiliary scripts module). Substring-A and -B each implement a different algorithm for consolidating the frequencies of word n-grams of different lengths (i.e. of different *n*). The process of frequency consolidation reduces the frequencies of substrings by the frequencies of the superstrings in which they are contained and an output list is produced showing the consolidated frequencies of all n-grams (see below for more on frequency consolidation). The auxiliary scripts module provides a number of additional functions related to the filtering of n-gram lists. The functions performed by this package will primarily be of interest to linguists and computational linguists working on formulaic language, multi-word expressions and other phraseological phenomena.




A. Overview of modules
-------------

### substring-A (algorithm with indexation)
This is a python script designed to work in conjunction with the [mwetoolkit] (http://mwetoolkit.sourceforge.net/PHITE.php). It takes as input a corpus-indexed list of n-grams (as produced by mwetoolkit) and then uses an exact, indexation-based algorithm to consolidate frequencies of overlapping n-grams, following [Altenberg and Eeg-Olofsson (1990: 16-17)](https://books.google.co.uk/books?id=ZjRO8cY0xwEC&printsec=frontcover&dq=Theory+and+Practice+in+Corpus+Linguistics+1990&hl=en&sa=X&ved=0ahUKEwiqvf6q_IDdAhUHJMAKHapOClAQ6AEIJzAA#v=onepage&q=Theory%20and%20Practice%20in%20Corpus%20Linguistics%201990&f=false).

The Substring-A algorithm is generally preferable to substring-B in cases where access is available to the source texts/corpora from which n-grams are to be extracted, and [mwetoolkit] (http://mwetoolkit.sourceforge.net/PHITE.php) can be used to extract n-grams from those texts.

### substring-B (algorithm without indexation)
This is a set of Unix shell scripts designed to work on a range of simpler n-gram list formats (where n-grams are not indexed to their occurrences in a specific corpus). Such lists are produced by the [NGramProcessor](http://buerki.github.io/ngramprocessor/), the [Ngram Statistics Package](http://www.d.umn.edu/~tpederse/nsp.html) and various other tools and sources such as [Google Books Ngrams](http://storage.googleapis.com/books/ngrams/books/datasetsv2.html). The algorithm implemented by substring-B is described in detail in [Buerki (2017)](https://doi.org/10.1007/978-3-319-69805-2_30).

If access to source texts/corpora is unavailable (only n-gram lists are available) or if the limitations of [mwetoolkit] (http://mwetoolkit.sourceforge.net/PHITE.php) mean that a different tool needs to be used for n-gram extraction, substring-B should be used. Access to source corpora may be unavailable in case of sources like [Google Books Ngrams](http://storage.googleapis.com/books/ngrams/books/datasetsv2.html), or because online corpus portals such as the [Sketch Engine](https://www.sketchengine.eu) are used that allow the creation of n-gram lists, but not the download of the full underlying corpus data. N-gram lists might require some formatting before processing, see the README file for substring-B.

### auxiliary scripts

Several auxiliary scripts are included in the SubString package which allow the further processing of n-gram lists after frequency consolidation. These scripts can be used after processing with substring-A or substring-B. See the TUTORIAL included in SubString for details on these scripts.


B. Frequency Consolidation
--------------------------

To illustrate frequency consolidation among n-grams of various lengths, let us assume we have as input the n-grams in (1)a. These will have been extracted from a corpus and their frequency of occurrence in the corpus is indicated by the number to the right of each n-gram.

The 4-gram 'have a lovely time' occurs with a frequency of 15. The 3-grams 'have a lovely' and 'a lovely time' occur 58 and 44 times respectively. 15 of those occurrences are, however, occurrences as part of the superstring 'have a lovely time'. To get the consolidated frequency of occurrence for 'have a lovely' and 'a lovely time' (i.e. the occurrences of these 3-grams on their own, NOT counting when they occur in a longer string), we therefore deduct the frequency of their superstring (15) from their own frequency. This results a consolidated frequency of 43 for 'have a lovely' (i.e. 58 minus 15) and 29 for 'a lovely time' (i.e. 44 minus 15), as shown in (1)b.

The remaining 2-grams ('have a', 'a lovely' and 'lovely time') are also substrings of 'have a lovely time' and therefore also need to have their frequency reduced by 15 (resulting in a frequency of 34692 for 'have a', 86 for 'a lovely' and 30 for 'lovely time'. In addition, 'have a' and 'a lovely' are substrings of 'have a lovely' and therefore the frequency of 'have a lovely' which is now 43, needs to be deducted from their frequencies. This results in a new frequency of 34649 for 'have a' and 43 for 'a lovely'. 'a lovely' and 'lovely time' are furthermore substrings of 'a lovely time' and consequently need to have their frequencies reduced by that of 'a lovely time' (i.e. by 29): the consolidated frequency of 'a lovely' is now 14, that of 'lovely time' is 1. The output of the frequency consolidation is shown in (1)b.

       (1)a                          (1)b
       
       have a lovely time  15        have a lovely time   15
       have a lovely       58        have a lovely        43
       a lovely time       44        a lovely time        29
       have a           34707        have a            34649
       a lovely           101        a lovely             14
       lovely time         45        lovely time           1

**Substring-A** implements an algorithm that only consolidates frequencies of n-grams that actually overlap in the source corpus. **Substring-B** has no access to this information and some assumptions are therefore made regarding what sequences are substrings of what other sequences without the possibility of checking this on the basis of indexation to a source text. While consolidations are coherent and accurate with regard to input lists in both algorithms, results of substring-A will inherently differ from those obtained from substring-B.

For more on frequency consolidation among n-grams see [Altenberg and Eeg-Olofsson (1990: 16-17)](https://books.google.co.uk/books?id=ZjRO8cY0xwEC&printsec=frontcover&dq=Theory+and+Practice+in+Corpus+Linguistics+1990&hl=en&sa=X&ved=0ahUKEwiqvf6q_IDdAhUHJMAKHapOClAQ6AEIJzAA#v=onepage&q=Theory%20and%20Practice%20in%20Corpus%20Linguistics%201990&f=false), [Buerki (2017)](https://doi.org/10.1007/978-3-319-69805-2_30) and [O'Donnell (2011)](
http://icame.uib.no/ij35/Matthew_Brook_ODonnell.pdf).


C. Dependencies
------
SubString should run on all systems that can run the bash shell. Substring-A additionally requires Python 3 and an installation of the [mwetoolkit 3](https://gitlab.com/mwetoolkit/mwetoolkit3). For efficient processing of larger amounts of data, substring-B requires bash 4 (although the software will automatically substitute a slower algorithm if only bash 3 is available).[^2]


[^2]: Most recent operating system versions have bash v. 4 installed as standard, but MacOS X has bash v. 3.2 installed as standard. Bash v. 4 can be installed using [MacPorts](http://www.macports.org), [Homebrew](http://brew.sh) or similar and then the new version would either need to be put in the directory `/bin` (replacing the old version) or the first line of the `substring-B.sh` script would need adjusting to point to the new version of bash (if installed via MacPorts, the new line would read `#!/opt/local/bin/bash` instead of `#!/usr/bin/env bash`).



D. Installation
------------

#### 1) Using the supplied installers:

Double-clickable installers are provided for OS X, Linux (with Gnome desktop) and Cygwin/Windows, and an installer script for all UNIX/Linux-like systems. These installers replace previous versions of the SubString package. It may be necessary to log out and log in again before the installation takes full effect.


###### Gnome Linux / Mac OSX

Inside the SubString directory, double-click on `Linux_installer` (for Ubuntu and other flavours with the Gnome desktop environment) or `OSX_installer` (OS X).


###### Any Unix/Linux-like system

Open a terminal window, drop the `install.sh` script (located inside the `bin` directory) onto the terminal window and press ENTER. This will start the interactive installation process.


###### Windows

The [Cygwin](cygwin.com) environment needs to be installed first. During the installation procedure for Cygwin, the optional 'bc' package from the 'maths' category will need to be installed as well as the 'python3' package from the 'Python' category. A guide on how to install Cygwin is found [here](http://x.cygwin.com/docs/ug/setup-cygwin-x-installing.html).

After Cygwin has been installed, double click on the `Cygwin_installer` or `Cygwin64_installer` (try both if one does not work) to start the installation process.

Especially under Windows systems, there can be problems if file and/or folder names contain spaces, so it is best to avoid locations on the system that have folder names with spaces.

#### 2) Manual installation / other flavours of Linux:

To install the substring-A module, copy `substring-A.py` (inside the `bin` directory) to the same bin directory as the other mwetoolkit files; `ft_txtcandidates.py` should be placed in the `bin/mwetk/filetype` directory of the mwetoolkit installation. The location should be in the users $PATH variable.

All other scripts in the `bin` folder should be placed in a location that is in the user's $PATH variable (or the location should be added to the $PATH variable) so they can be called from the command line. A good place to put the scripts might be `/usr/local/bin` or `~/bin`.

To uninstall, open a terminal window then drop the `install.sh` script (located inside the `bin` directory) on to the terminal window, then type ` -u` so that the line ends in `install.sh -u` and press ENTER. Alternatively, manually delete the relevant files from their installation locations as indicated above.



E. Basic Operation
------------------

The frequency consolidation modules A and B can be accessed in interactive mode by double-clicking on the SubString icon (If an icon was generated on the desktop or in the applications folder during installation). Alternatively, it can be started from a terminal by typing `substring.sh` and pressing ENTER. In non-interactive mode, the three modules of SubString are used as shown below. See the TUTORIAL for examples and more information.

#### substring-A module

This module is invoked from the command line like so `substring-A.py [OPTIONS] INPUT_FILE.xml`; options are displayed by running `substring-A.py -h` and input files must be in mwetoolkit's XML format.


#### substring-B module

Substring-B is probably best used interactively, but can be invoked from the command line like so `substring-B.sh [OPTIONS] [-u uncut_list]+ INPUT_FILE.txt+`, where `INPUT_FILE.txt+` is two or more n-gram lists of different length (use `-h` or the TUTORIAL for usage details).

N-gram lists processed by the B module must be formatted in the correct format (see `test_data-B` folder for examples). N-gram lists produced by the [N-Gram Processor](http://buerki.github.io/ngramprocessor) are suitable for direct input; if n-gram lists were produced by a different programme they might need reformatting. 


#### auxiliary scripts

* `cutoff.sh -f N FILE+` to apply a frequency cutoff to an n-gram list
* `length-adjust.sh -c N [-OPTIONS] FILE+` see TUTORIAL for details
* `en-filter.sh [-OPTIONS] FILE+` to apply a structural filter to n-gram lists
* `random_lines.sh [-OPTIONS] FILE` to extract random lines from a list
* `TP-filter.sh [-OPTIONS] FILE` to assess the precision of a filtered list of n-grams
* `listconv.sh [-OPTIONS] FILE[S]` to convert n-gram lists between different formats

These scripts accept n-gram lists in the format of the [N-Gram Processor](http://buerki.github.io/ngramprocessor) which is also the format of substring-B and can be specified as output format to substring-A by specifying the option `--to NGP`.

See the TUTORIAL file included in the SubString directory for examples and more information on all three modules.

F. Known Issues
---------------

On OSX it's been reported that on some systems, clicking on the SubString icon in the applications folder or desktop brings up the following error: 'SubString can’t be opened because Sandbox is not allowed to open documents in Terminal.' If this happens, delete the SubString app that brings up this message and instead use the file `SubString` (with the icon of a black terminal window in a white frame) to launch SubString interactively. This file is located in the `bin` directory within the SubString directory.

Issues can be raised at [http://github.com/buerki/SubString/issues](http://github.com/buerki/SubString/issues), but no active support can be provided for this software.



G. Copyright, licensing, download
---------------------------------

Substring is (c) 2016-2018 Cardiff University, 2011-2015 Andreas Buerki, licensed under the EUPL V.1.1. (the European Union Public Licence) which is an open-source licence (see the EUPL.pdf file for the full licence).

The project resides at [http://buerki.github.com/SubString/](http://buerki.github.com/SubString/) and new versions will be posted there. To be notified of new releases, go to https://github.com/buerki/SubString, click on the 'Watch' button and sign in.

H. Warning
----------

As article 7 of the EUPL states, the SubString is a work in progress, which is continuously improved. It is not a finished work and may therefore contain defects or “bugs” inherent to this type of software development.

release_notes.txt
release notes v. 1.2
********************

substring-B.sh
- fixed an issue introduced in version 1.1.2, where the order of consolidation was mixed up for n-grams of n > 9.
- extended limit of 12 uncut lists to 29 uncut lists.

release notes v. 1.1.2
**********************

substring.sh
- enabled the consolidation of n-grams of lengths up to n = 30.


release notes v. 1.1.1
**********************

substring.sh
- fixed an issues with a failure to copy consolidated list to output directory when path contains spaces

substring-B.sh
- fixed an issue that stopped correct processing of lists if no bigram list was available. 3-gram list is still required.


release notes v. 1.1
********************

install.sh & substring-A.py
- changed name of libraries folder from 'libs' to 'mwetk' in line with changes in the latest version of the mwetoolkit

cutoff.sh
- added option that allows processing of file in Google n-gram format


release notes v. 1.0
********************

A new, modular architecture was introduced, splitting SubString into three modules. The main algorithm of SubString up to version 0.9.9.2 was retained as one of the modules and a new module (substring-A.py) added that implements a frequency consolidation algorithm that makes use of mwetoolkit's indexing of n-grams. The auxiliary scripts were retained as the third module.

substring.sh
- adjusted to the modular architecture

TP-filter, cutoff.sh, random_lines.sh, length-adjust.sh
- changed handling of filename extensions so that extensions are preserved correctly

substring-processor.sh
- renamed substring-B.sh

newly added:
- substring-A.py
- libs/filetype/ft_ngp.py & ft_nsp.py
- xml_list_to_NGP.py
- TUTORIAL.md
- plaintext_list.xsl


release notes v. 0.9.9.2
************************

substring-processor.sh
- raised limit of number of uncut lists to be incorporated in the consolidation to 12.

TP-filter.sh
- added option to rate on a scale from 1 to 6 rather than a binary T/F distinction.


release notes v. 0.9.9.1
************************

substring-processor.sh
- fixed an issue with the progress indicator not advancing during the preparatory stage


release notes v. 0.9.9
**********************

substring.sh
- added a menu-based interface
- moved main processing code to new substring-processor.sh

Added the following auxiliary scripts used by substring.sh
- consolidate.sh
- en-filter.sh
- random_lines.sh
- TP-filter.sh
- substring-processor.sh

listconv.sh
- script was retired as the NGP package now provides suitably formatted input files.

Removed awk commands from all scripts due to compatibility issues with Cygwin in certain configurations

Adjusted the following to reflect above changes
- install.sh
- README.pdf/README.md:
- test_data


release notes v. 0.9.7
**********************

Added double-clickable installers for Cygwin and a generic installer script:
- Cygwin_installer.lnk
- Cygwin64_installer.lnk
- install.sh

README.pdf/README.md:
- adjusted for changes

release notes v. 0.9.6
**********************

Added double-clickable installer for OS X and Linux:
- OSX_installer.command
- linux_installer.desktop

README.pdf/README.md:
- adjusted for changes

release notes v. 0.9.5
**********************

substring.sh:
- prep stage processing now makes use of Bash 4's associative arrays – this 
  enables the processing of far larger amounts of data (previous algorythm 
  still works if no bash 4 is detected)
- fixed bug introduced in 0.9.4 which resulted in hung processing if the -d
  option was invoked
- removed trailing tabs from output lists inadvertently introduced in 0.9.4
- a few other efficiency improvements

README.txt/README.md:
- adjusted for changes

test_data:
- updated gold lists to reflect changes to substring.sh

release notes v. 0.9.4
**********************

README.txt/README.md:
- adjusted for changes
- description for length-adjust.sh added

substring.sh:
- consolidation down to single words enabled (before it was down to bigrams)
- fixed some linux-specific issues with coreutil's cut
- fixed an issue where a warning was issued about unexpected format in lists
  had n-grams with underscores in them
- efficiency improved by reducing the number of I/Os
- improved accuracy during imports at preparatory stage

cutoff.sh:
- improved efficiency
- introduced -i option for improved naming of cutoff lists

length-adjust.sh (new):
- newly added script to adjust lengths of certain n-grams after consolidation

listconv.sh:
- fixed bug in -n option
- expanded number of input list formats handled

test_data:
- updated gold lists to reflect changes to substring.sh
- new data in example3 that includes word lists
- new example4

release notes v. 0.9.1
**********************

README.txt: 
- adjusted for new home page url
- adjusted to reflect changes below

cutoff.sh, listconv.sh, substring.sh:
- now operate with flexible n-gram constituent separators for compatibility
  with the N-Gram Processor (http://buerki.github.io/ngramprocessor)
  
test_data
- added GOLD standard lists

release notes v. 0.9
********************

README.txt: 
- adjusted recommended installation directory 

cutoff.sh: 
- added -n N option which restricts the cutoff to n-grams of size N. This
  is useful if a list with different lengths of n-grams (such as after 
  consolidation with substring.sh) needs to have length-specific cutoffs
  applied. The -n option only processes lists of formats 'n<>gram<> 370' or
  'n<>gram<>	370	258'
- expanded help available in -h option

substring.sh:
- added an identifyer line to the neg-freq.lst file which logs negative
  frequencies
- renamed the -m option to -n option as this is a more easily remembered name
  for the option to retain negative frequency strings and 0-frequency strings in
  the output lists
- added -z option to include sequences with zero-frequency (but not with
  negative frequency) in the final output list 
- adjusted help (-h) accordingly

release notes v. 0.8.7
**********************
substring.sh:

src/cutoff.sh
#!/bin/bash -
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:"$HOME/bin"" # needed for Cygwin
##############################################################################
# cutoff.sh
copyright="Copyright (c) 2016 Cardiff University, 2011-2014 Andreas Buerki"
# licensed under the EUPL V.1.1.
version="0.9.9"
####
# DESCRRIPTION: enforces frequency cutoffs in n-gram lists
# SYNOPSIS: cutoff.sh [-f 'regex'][-d 'regex'] [-a 'regex'] FILE(S)
# OPTIONS:	see -h
##############################################################################
# History
# date			change
# 13/01/2011	made output naming more consistent (no trailing dot if no doc
#				cutoff is specified as appeared sporadically before) by ad-
#				justing the method to determine list structure (tab count).
# 04/02/2011	added -v option, fixed recognition of rank stat freq lists
# 10/02/2011	fixed -a option
# 30/04/2011	errors channelled to strderr
# 22/12/2011	added -V option
# 07 Jan 2012	adjusted verbose behaviour to list the lists that will be
#				processed
# 11 Jan 2012	added add_to_name function for output filenames
# (0.8.7)
# 17 Feb 2012	added -n option and expanded information in -h option
# (0.8.8)	
# 25 Nov 2013	made n-gram separators flexible; can not detect · _ or <> as
# (0.8.9)		as valid constituent separators	or whatever is supplied in -p
# 01 Sep 2014	made a few optimisation changes (replaced all cat | grep)
# 07 Sep 2014   added -i option
# (0.9.1)
# 09 Jan 2016	created an easier way to specify cutoff frequencies using -f
# (0.9.9)		and -d (previous way to operate -f/d shifted to -F/-D options) 

# define functions
help ( ) {
	echo "
Usage: $(basename $0)  [-f 'regex'][-d 'regex'] FILE(S)
Example: $(basename $0)  -f 30 -d 1 list1.txt list2.txt
Options: -f N where 'N' is the minimum frequency allowed (a number between 1 and 101)
         -F requires as argument a regular expression matching
            the n-gram frequencies that should be cut out,
            to cut out frequencies 1-15 the pattern '([0-9]|[1][0-5])'
            should be given as argument (mind the quotes and parentheses)
         -d analog to -f, but applies to document frequencies
         -D analog to -F, but applies to document frequencies, i.e.
            to cut all n-grams that occur in only one document, the
            argument '1' should be given (no quotes necessary as this
            does not contain meta characters)
         -i intelligible name: uses the last number of the freqency cutoff
            regular expression (supplied via -f), rather than the whole
            expression, in naming the output file. For this to work a
            number (not range of numbers) must be provided as the last
            element in the regular expression, i.e. for all below 11 it
            would be '([1-9]|10|11)' rather than '([1-9]|1[01])'.
         -a analog to -f and -d, but applied to the score of association
            measures (if listed). use only integers, it will cut the
            integers AND following decimals,
            i.e. if AM score [0-5] is entered, 0.0000 to 5.9999 is cut)
         -n N restricts the cutoff to n-grams of size N. This is useful if
            a list with different lengths of n-grams needs to have length-
            specific cutoffs applied.
         -p SEP provide the separator character used to separate elements of
            n-grams in the list (if the script cannot guess it).
NOTES:
input files must contain data in one of these formats:

	1	'n·gram· 370'
		(i.e. n-gram of any size, with tab delimited frequency count, w/o
		trailing space)

	2	'n·gram·	370	258'
 		(i.e. n-gram of any size, tab delimited without trailing space,
		first number is freq., second doc count)

	3	'n·gram·	370	258	T|F'
 		(i.e. n-gram of any size, tab delimited without trailing space,
		first number is freq., second doc count, then TP/TP designation)

	4	'n·gram·	1	1128.4296	21'
		(i.e. as immediately below, but without document count)

	5	'n·gram·	1	1128.4296	21	1'
		(i.e. n-gram of any size, tab or space delimited rank, association 
		measure, frequency and document frequency, without trailing space)

The script detects these formats automatically and deals with each accoringly.
This script implements n-gram frequency and, if applicable, document 
frequency cut-offs. It cuts out items on the n-gram lists as specified
by the user-supplied regex string.
It appends '.cut.X[.X]' to the cut list (where X is the cut-off regex
for frequency and doc); it leaves the uncut list in place
"
}
Usage ( ) {
	echo "
Usage: $(basename $0)  [-f N][-d N] [-n N] [-v] FILE[S]
use -h for help
"
}
# define add_to_name function
add_to_name ( ) {
#####
# this function checks if a file name (given as argument) exists and
# if so appends a number at the end of the name so as to avoid overwriting 
# existing files of the name as in the argument or any with the same name as the 
# argument plus an incremented number count appended.
####
count=
if [ -a $1 ]; then
	add=-
	count=1
	while [ -a $1-$count ]
		do
		(( count += 1 ))
		done
else
	count=
	add=
fi
output_filename=$(echo "$1$add$count")
}
# define regex_maker function
regex_maker ( ) {
#deduct 1 to get the cut-frequency
(( cut_freq -= 1 ))
# work out the correct regex for cutting
num_of_digits=$(( $(echo $cut_freq | wc -c) - 1 ))
if [ $num_of_digits -eq 1 ]; then
	export cut_regex="[0-$cut_freq]|$cut_freq"
elif [ $num_of_digits -eq 2 ]; then
	first_digit=$(sed 's/\(^[[:digit:]]\).*/\1/g' <<< $cut_freq)
	first_dm1=$(( $first_digit - 1 ))
	second_digit=$(sed 's/^.\([[:digit:]]\).*/\1/g' <<< $cut_freq)
	export cut_regex="[0-9]|[0-$first_dm1][0-9]|$first_digit[0-$second_digit]|$cut_freq"
elif [ $num_of_digits -eq 3 ]; then
	export cut_regex="[1-9]|[1-8][0-9]|9[0-9]|100"
	if [ $cut_freq -gt 100 ]; then
		echo "The value entered is greater than 101. The value 101 will be used instead."
	fi 
fi
}
########################## end defining functions
# analyse options
while getopts ha:F:f:D:d:in:p:vV opt
do
	case $opt	in
	h)	help
		exit 0
		;;
	a)	score=$OPTARG
		a=given
		;;
	f)	cut_freq=$OPTARG
		f=given
		regex_maker
		nfreq="($cut_regex)"
		;;
	F)	nfreq=$OPTARG
		f=given
		;;
	d)	cut_freq=$OPTARG
		d=given
		regex_maker
		doc="($cut_regex)"
		;;
	D)	doc=$OPTARG
		d=given
		;;
	i)	intelligible=true
		;;
	n)	nsize_restriction=$OPTARG
		;;
	p)	separator=$OPTARG
		;;
	v)	verbose=true
		;;
	V)	echo "$(basename $0)	-	version $version"
		echo "$copyright"
		echo "licensed under the EUPL V.1.1"
		echo "written by Andreas Buerki"
		exit 0
		;;
	esac
done
shift $((OPTIND -1))
# check if any arguments are supplied at all
if [ $# = 0 ]
	then
		Usage
		exit 1
fi
if [ "$verbose" == "true" ]; then
	echo "the following files will be processed"
	echo "$(echo $@ | sed 's/ /\
/g')"
	if [ -n "$nsize_restriction" ]; then
		echo "only n-grams of size $nsize_restriction will be processed"
	fi
	echo " "
fi
# check if arguments exist
for list; do
	if [ -s $list ]; then
		:
	else
		echo "$list not found or empty" >&2
		exit 1
	fi
done
# if -n option is active,
# create scratch directory where temp files can be moved about
if [ -n "$nsize_restriction" ]; then
	SCRATCHDIR=$(mktemp -dt cutoffXXX)
	# if mktemp fails, use a different method to create the scratchdir
	if [ "$SCRATCHDIR" == "" ] ; then
		mkdir ${TMPDIR-/tmp/}cutoff.$$
		SCRATCHDIR=${TMPDIR-/tmp/}cutoff.$$
	fi
fi
for arg in $@
		do
			if [ "$verbose" == "true" ]; then
				echo "processing file $arg ..."
			fi
			# eliminate total n-gram count if present
			if [ "$(head -1 "$arg" | grep '^[[:digit:]]*.$')" ] || [ "$(head -1 "$arg" | grep '^[[:digit:]]*$')" ]; then
				sed 1d "$arg" > "$arg."
				mv "$arg." "$arg"
				restore=true
			fi
			# set separator and
			# set nsize variable
			# checking input list to derive separator and nsize
			# check if -p option is active and if so, use that separator
			# check separator for current list
			if [ "$separator" ]; then
				if [ "$(head -1 $list | grep "$separator")" ]; then
					:
				# if the separator given is not found and list is not empty,
				# produce error
				elif [ -s $list ]; then
					echo "separator $separator not found in $(head -1 $list) of file $list" >&2
					exit 1
				fi
			elif [ "$(head -1 $list | grep '<>')" ]; then
				separator='<>'
				short_sep='<'
			elif [ "$(head -1 $list | grep '·')" ]; then
				separator="·"
				short_sep="·"
			else
				echo "unknown separator in $(head -1 $list) of file $list" >&2
				exit 1
			fi
			# extract n of n-gram list
			nsize=$(head -1 "$arg" |  tr -dc "$short_sep" | wc -c | sed 's/ //'g)
			# this counts binary separator characters as two, so needs adjusting
			if [ "$short_sep" == "·" ]; then
				nsize=$(( nsize / 2 ))
			fi
			if [ "$verbose" == "true" ]; then
				echo "separator is $separator"
				echo "nsize is $nsize"
			fi
			# check for data structure in input files and treat accordingly
			# 1 tab -> case 1: n·gram·	freq
			# 2 tabs -> case 2: n·gram·	freq	doc
			# 3 tabs, T|F as last character -> case 3: n·gram· freq	doc	T/F
			# 3 tabs, not T|F as last char. -> case 4: n·gram· rank stats freq
			# 4 tabs -> case 5: n·gram·	rank	stats	freq	doc
			# case 1
			if [ $(head -1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 1 ] ; then # n-gram freq
				if [ "$verbose" == "true" ]; then
					echo "1tab recognised"
				fi
				# check if impossible cutoffs were specified
				if [ "$d" = "given" ]; then
					echo "the list $arg does not seem to contain document frequency information: $(head -1 $arg)" >&2
				elif [ "$a" = "given" ]; then
					echo "this list does not seem to contain score information: $(head -1 $arg)" >&2
					exit 1
				fi
				if [ -z "$nsize_restriction" ]; then
					# make name for output file
					if [ "$intelligible" == true ]; then
						add_to_name $arg.cut.$(echo $nfreq | sed 's/.*|\([[:digit:]]*\))/\1/')
					else
						add_to_name $arg.cut.$nfreq
					fi
					# execute desired cutoff
					grep -E -v "$separator	$nfreq$|^[0-9]*$" $arg > $output_filename
				else
					# make name for output file
					add_to_name $arg.cut.$nfreq.$nsize_restriction-grams
					# execute desired cutoff
					for line in $(sed 's/	/•/g' $arg); do
						if [ $nsize -eq $nsize_restriction ] && [ -n "$(echo $line | egrep "$separator•$nfreq$")" ]; then
							:
						else
							echo $line | sed 's/•/	/g' >> $SCRATCHDIR/tmp.lst
						fi
					done
					grep -E -v "^[0-9]*$" $SCRATCHDIR/tmp.lst > $output_filename
				fi
			# case 2	
			elif [ $(head -1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 2 ] ; then # n-gram	freq doc
				if [ "$verbose" == "true" ]; then
					echo "2tabs recognised"
				fi
				# check if impossible cutoffs were specified
				if [ "$a" = "given" ]; then
					echo "this list does not seem to contain score information: $(head -1 $arg)" >&2
					exit 1
				fi
				# check if docment cutoff was specified
				if [ "$d" = "given" ]; then
					if [ -z "$nsize_restriction" ]; then
						# make name for output file
						if [ "$intelligible" == true ]; then
						add_to_name $arg.cut.$(echo $nfreq | sed 's/.*\|\([[:digit:]]*\))/\1/').$doc
					else
						add_to_name $arg.cut.$nfreq.$doc
					fi
						# execute desired cutoffs
						grep -E -v "$separator	$nfreq	|$separator	[0-9]*	$doc$|^[0-9]*$" $arg > $output_filename
					else
						# execute desired cutoff
						for line in $(sed 's/	/•/g' $arg); do
							if [ $nsize -eq $nsize_restriction ] && [ -n "$(echo $line | sed 's/•/	/g' | egrep "$separator	$nfreq	|$separator	[0-9]*	$doc$|^[0-9]*$")" ]; then
								:
							else
								echo $line | sed 's/•/	/g' >> $SCRATCHDIR/tmp.lst
							fi
						done
						# make name for output file
						add_to_name $arg.cut.$nfreq.$doc.$nsize_restriction-grams
						# move list into place
						mv $SCRATCHDIR/tmp.lst $output_filename
					fi	
				else
					# no doc-cutoff specified
					if [ -z "$nsize_restriction" ]; then					
						# make name for output file
						if [ "$intelligible" == true ]; then
						add_to_name $arg.cut.$(echo $nfreq | sed 's/.*\|\([[:digit:]]*\))/\1/')
					else
						add_to_name $arg.cut.$nfreq
					fi
						# execute desired cutoffs
						grep -E -v "$separator	$nfreq	|^[0-9]*$" $arg > $output_filename
					else
						# make name for output file
						add_to_name $arg.cut.$nfreq.$nsize_restriction-grams
						# execute desired cutoff
						for line in $(sed 's/	/•/g' $arg); do
							if [ $nsize -eq $nsize_restriction ] && [ -n "$(echo $line | sed 's/•/	/g' | egrep "$separator	$nfreq	|^[0-9]*$")" ]; then
								:
							else
								echo $line | sed 's/•/	/g' >> $SCRATCHDIR/tmp.lst
							fi
						done
						mv $SCRATCHDIR/tmp.lst $output_filename
					fi
				fi
			# if tab-delimited with 3 tabs	
			elif [ $(tail -n 1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 3 ] ; then
				if [ "$verbose" == "true" ]; then
					echo "3tabs recognised"
				fi
				if [ -n "$nsize_restriction" ]; then
					echo "the -n option is not implemented for this type of list" >&2
					exit 1
				fi
				if [ -n "$(tail -n 1 $arg | grep [TF]$ | cut -f 1)" ]; then
				# n-gram freq doc T|F
				if [ "$verbose" == "true" ]; then
					echo "T/F list recognised"
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$doc
				grep -E -v "$separator	$nfreq	|$separator	[0-9]*	$doc	[TF]$|^[0-9]*$" $arg > $output_filename
				else # n-gram rank stats freq
				if [ "$verbose" == "true" ]; then
					echo "rank stats freq list recognised"
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$score
				grep -E -v "$separator	[0-9]*	[0-9]*\.[0-9]*	$nfreq$|$separator	[0-9]*	$score\.[0-9]*" $arg > $output_filename
				fi
			elif [ $(tail -n 1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 3 ] ; then # if space-delimited with 3 spaces
				if [ "$verbose" == "true" ]; then
					echo "3space recognised"
				fi
				if [ -n "$nsize_restriction" ]; then
					echo "the -n option is not implemented for this type of list" >&2
					exit 1
				fi
				if [ -n $(tail -n 1 $arg | grep "[TF]$" | cut -f 1) ]; then
				# n-gram freq doc T|F
				if [ "$verbose" == "true" ]; then
					echo "n-gram freq doc T|F list recognised"
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$doc
				grep -E -v "$separator $nfreq |$separator [0-9]* $doc	[TF]$|^[0-9]*$" $arg > $output_filename
				else # n-gram rank stats freq
				if [ "$verbose" == "true" ]; then
					echo "n-gram rank stats freq list recognised"
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$score
				grep -E -v "$separator [0-9]* [0-9]*\.[0-9]* $nfreq$|$separator [0-9]* $score\.[0-9]*" $arg > $output_filename
				fi
			elif [ $(head -1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 4 ] ; then # n-gram rank stats freq doc
				if [ "$verbose" == "true" ]; then
					echo "n-gram rank stats freq doc list recognised"
				fi
				if [ -n "$nsize_restriction" ]; then
					echo "the -n option is not implemented for this type of list" >&2
					exit 1
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$doc.$score
				grep -E -v "$separator	[0-9]*	[0-9]*\.[0-9]*	$nfreq	|$separator	[0-9]*	[0-9]*\.[0-9]*	[0-9]*	$doc$|$separator	[0-9]*	$score\.[0-9]*" $arg > $output_filename
			elif [ $(head -1 $arg | tr -dc '	' | wc -c | sed 's/ //'g) -eq 4 ] ; then # n-gram rank stats freq doc, space delimited
				if [ "$verbose" == "true" ]; then
					echo "n-gram rank stats freq doc, space delimited list recognised"
				fi
				
				if [ -n "$nsize_restriction" ]; then
					echo "the -n option is not implemented for this type of list" >&2
					exit 1
				fi
				# make name for output file
				add_to_name $arg.cut.$nfreq.$doc.$score
				grep -E -v "$separator [0-9]* [0-9]*\.[0-9]* $nfreq |$separator [0-9]* [0-9]*\.[0-9]* [0-9]* $doc$|$separator [0-9]* $score\.[0-9]*" $arg > $output_filename
			else
				echo "input list format not recognised: $(head -n 1 $arg)" >&2
				exit 1
			fi
		done
# tidy up
if [ -n "$nsize_restriction" ]; then
	rm -r $SCRATCHDIR
fi